server/calls_targets.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"time"
+
+	promModel "github.com/prometheus/common/model"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+func resolveURL(u string, timeout time.Duration) ([]net.IP, string, error) {
+	parsed, err := url.Parse(u)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to parse url: %w", err)
+	}
+
+	host, port, err := net.SplitHostPort(parsed.Host)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to split host/port: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	ips, err := net.DefaultResolver.LookupIP(ctx, "ip4", host)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to lookup ips: %w", err)
+	}
+
+	return ips, port, nil
+}
+
+func (p *Plugin) generateCallsTargets(appCfg *model.Config, host, port string, nodes []*model.ClusterDiscovery) ([]promModel.LabelSet, error) {
+	// First, figure out if Calls is running. If so, add the plugin metrics endpoint to the targets.
+	// Also, check if the external RTCD service is configured, in which case add its endpoints to targets.
+	status, err := p.API.GetPluginStatus(callsPluginID)
+	if err != nil {
+		return nil, fmt.Errorf("generateCallsTargets: failed to get calls plugin status: %w", err)
+	}
+
+	if status.State != model.PluginStateRunning {
+		p.API.LogDebug("generateCallsTargets: calls plugin is not running")
+		return nil, nil
+	}
+
+	p.API.LogDebug("calls plugin running, generating targets")
+
+	var targets []promModel.LabelSet
+	if len(nodes) < 2 {
+		targets = []promModel.LabelSet{
+			{
+				promModel.AddressLabel:     promModel.LabelValue(net.JoinHostPort(host, port)),
+				promModel.MetricsPathLabel: promModel.LabelValue(fmt.Sprintf("/plugins/%s/metrics", callsPluginID)),
+			},
+		}
```

Comment:
I don't know what the best practice here is. Maybe there's a better grouping we could use because, as it stands, I am adding all the targets to the same slice along with MM ones. Also, the Calls plugin target technically has the same address as base MM, the only difference being the additional path. Not sure if that would cause issues given some process metrics have the same name.

Reply:
Good point, we probably be okay with the label configuration which accompanies to the target configuration. Did you had any chance to test it?

Reply:
I am still to use the dump, so I need to verify how it's looking once loaded. Will report back.

Reply:
As long as we distinguish targets related to calls then there shouldn't be any conflict since the endpoint/address would be different. So I am enforcing the job label to be `calls` which should do the trick.